File: tests/07_phases_hlt.dg/success_collapse_02.cpp
/*
<testinfo>
test_generator="config/mercurium-hlt run"
</testinfo>
*/

#include <iostream>
#include <cassert>

int main()
{
    // cppreference.com asserts:
    //      "C++ Standard guarantees that 1 == sizeof(char) <= sizeof(short) <= sizeof(int) <= sizeof(long) <= sizeof(long long)."

    // result (effective) iteration space (product) will be > 2^(8-1) - 1, in fact it is exactly 2^7

    char lb_i = 0;
    char lb_j = 0;
    char lb_k = 0;

    char ub_i = 4;
    char ub_j = 4;
    char ub_k = 8;

    char s_i = 1;
    char s_j = 1;
    char s_k = 1;

    unsigned int result = 0;

    // Collapsed loops

    #pragma hlt collapse(3)
    for (char i = lb_i; i < ub_i; i += s_i)
    for (char j = lb_j; j < ub_j; j += s_j)
    for (char k = lb_k; k < ub_k; k += s_k)
    {
        result++;
    }

    assert(result == (1 << 7));
    std::cout << "result: " << result << std::endl;
}

File: tests/07_phases_hlt.dg/success_collapse_02.f90
! <testinfo>
! test_generator="config/mercurium-hlt run"
! </testinfo>

PROGRAM P
    IMPLICIT NONE
    INTEGER(1) :: I, J, K
    INTEGER(1) :: LB_I, LB_J, LB_K
    INTEGER(1) :: UB_I, UB_J, UB_K
    INTEGER(1) :: S_I, S_J, S_K
    INTEGER(4) :: RESULT

    ! Total (effective) iteration space (product) will be > 2^(8-1) - 1, in fact it is exactly 2^7

    LB_I = 1
    LB_J = 1
    LB_K = 1

    UB_I = 4
    UB_J = 4
    UB_K = 8

    S_I = 1
    S_J = 1
    S_K = 1

    RESULT = 0

    ! Collapsed loops

    !$HLT COLLAPSE(3)
    DO I = LB_I, UB_I, S_I
        DO J = LB_J, UB_J, S_J
            DO K = LB_K, UB_K, S_K
                RESULT = RESULT + 1
            END DO
        END DO
    END DO
    !$HLT END COLLAPSE

    IF (RESULT .NE. 2**7) THEN
        WRITE(0, *) "Assertion `RESULT .NE. 2**7' failed."
        STOP
    END IF
END PROGRAM
